R/extrapolators.R

```diff
@@ -0,0 +1,86 @@
+# extrapolators.R
+# helper extrapolator functions for use in base year changes
+
+#' extrapolate_constant
+#'
+#'  computes the mean of last n values and uses this constant value to fill in
+#'  missing years at the end of the time series
+#'
+#' @param x Vector of values with NA's to be filled in via constant extrapolation
+#' of the mean of last n non-NA values.
+#' @param n Number of non-NA values to be averaged to provide the filler value.
+#' @details Computes the mean of last n non-NA values of input vector x
+#' and uses this constant value to fill in NA values in x.
+#' @return Vector with all NA values replaced with the specified mean.
+#' @importFrom assertthat assert_that is.scalar
+#' @author ACS June 2019
+extrapolate_constant <- function(x, n=1){
+
+  assert_that(is.numeric(x))
+  assert_that(is.scalar(n))
+
+  meanval <- mean(last_n(x,n))
+
+  x[is.na(x)] <- meanval
```

Comment:
This will obviously fill in _all_ NA values, not just at the end of the time series. Is this OK? 🤔

Reply:
Good catch - that's what will happen in both implementations (vector and tibble) and should definitely be corrected.

Reply:
This is a pretty important question.  Certainly we don't want to "interpolate" with the extrapolator.  I mean we could tell users to do any interpolations before extrapolations but that seems less than ideal.

What about a case for instance the dataset coming in was expected to be filled up to say 2010 and we want to extrapolate out to 2015 but region X was actually missing a value in 2010.  These methods would happily extrapolate the 2010 value as well in region X.  Maybe that is fine as well and we just make clear to users that if they expect values complete data up to year Y they should ensure that before passing it into the extrapolators.

Reply:
The new implementation of the extrapolator would handle this situation like this: 
The data set stops at 2010, but region X has an NA value in 2010. We want to do constant extrapolation to 2015. 

Assuming we grouped by region, everyone but X will behave the way we expect. 
As is, the extrapolator will take the 2009 value and fill it in for 2011-2015. The NA value in 2010 will be left alone. Endogenously determining how many tail end NA values are in a vector was proving frustrating, so I decided to just make an argument for that.

Reply:
Is this (filling in region X 2010 NA value) definitely a problem? As noted in previous comments, if we can allow this, it makes things much simpler (no need to calculate `numMissing`...).

Reply:
As I think about it, in our first go-round it won't be a problem because 2010 is our final calibration year, so by definition it is not missing.

Maybe we are just clear that this is what happens.  I think in future development the idea that region X would have to fill 2010 wouldn't be as strange as there wouldn't be this preconceived notion that we are only "extending to 2015 for BYU".